### PR TITLE
new IncrementalBuild to support building only modules provided as arg

### PR DESCRIPTION
**Pull request by rpatali**
this is more or less restoring of the original implementation
@jacobgreenleaf removed in PR #563. We didn't have use case for this
code to exists but our build times are getting longer and we need to way
to reduce the build time.

### CODE REVIEW

File: codegen/module.go [hunk 1]
```diff
@@ -1022,6 +1022,134 @@ func (system *ModuleSystem) getSpec(instance *ModuleInstance) (interface{}, bool
...
+					classInstance.ClassName,
+					classInstance.ClassType,
+				)
+				continue
+			}
+
+			toBeBuiltModules[classInstance.ClassName][classInstance.InstanceName] = classInstance
+		}
+
+		// Collect things of the same class that depend on us
```
Comment (jacobgreenleaf) [inline codegen/module.go]: This comment is inaccurate. You are collecting *all* things that depend (directly or indirectly) on something in `toBeBuiltModules`.

File: codegen/module.go [hunk 2]
```diff
@@ -1022,6 +1022,134 @@ func (system *ModuleSystem) getSpec(instance *ModuleInstance) (interface{}, bool
...
+		for _, classInstance := range classInstances {
+			// classInstance needs to be built if any of the dependencies that are to be built is in the recursive
+			// dependency tree of classInstance.
+
+			for _, className := range system.classOrder {
```
Comment (jacobgreenleaf) [inline codegen/module.go]: You can simplify this by moving this loop into a method like `getAllAncestors(*ModuleInstance, map[string][]*ModuleInstance)`, maybe even defining it as a method on the `map[string][]*ModuleInstance`.

File: codegen/module.go [hunk 3]
```diff
@@ -1022,6 +1022,134 @@ func (system *ModuleSystem) getSpec(instance *ModuleInstance) (interface{}, bool
 	return spec, true, nil
 }
 
+func (system *ModuleSystem) tryResolveIncrementalBuild(
+	instances []ModuleDependency,
+	resolvedModules map[string][]*ModuleInstance,
+) (map[string][]*ModuleInstance, error) {
+
+	toBeBuiltModules := make(map[string]map[string]*ModuleInstance, 0)
```
Comment (jacobgreenleaf) [inline codegen/module.go]: nit: rename `toBeBuiltModules` to `toBeBuiltModulesUnique` to note why it's a `map[string]map[string]*ModuleInstance`.

File: codegen/module.go [hunk 4]
```diff
@@ -1022,6 +1022,134 @@ func (system *ModuleSystem) getSpec(instance *ModuleInstance) (interface{}, bool
...
+					"Skipping generation of %q %q class of type %q "+
+						"as not needed for incremental build\n",
+					classInstance.InstanceName,
+					classInstance.ClassName,
+					classInstance.ClassType,
+				)
+				continue
+			}
+
+			toBeBuiltModules[classInstance.ClassName][classInstance.InstanceName] = classInstance
```
[the remark this answers is not in the document]
Comment (rpatali) [reply codegen/module.go]: I can definitely move this statement in the `if` block of the inner `for` loop but i think i still want to break that loop when a match is found. `found` is just an easier way for me to print that message although that can be derived too, i prefer this.

File: codegen/module.go [hunk 5]
```diff
@@ -1022,6 +1022,141 @@ func (system *ModuleSystem) getSpec(instance *ModuleInstance) (interface{}, bool
...
+	commitChange bool,
+) (map[string][]*ModuleInstance, error) {
+	if len(instances) == 0 {
+		fmt.Println("Skipping build since no module dependency is provided")
+		return make(map[string][]*ModuleInstance), nil
+	}
+
+	toBeBuiltModules, err := system.collectTransitiveDependencies(instances, resolvedModules)
+	if err != nil {
+		// if incrementalBuild fails, perform a full build.
```
Comment (rpatali) [inline codegen/module.go]: infof falling back

**Pull request merged**